ui/details-panel/AutoclassifyTab.jsx

```diff
@@ -0,0 +1,535 @@
+
+    if (job.id) {
+      const matcherResp = await fetch(getApiUrl("/matcher/"));
+      const matcherData = await matcherResp.json();
+      const errorMatchers = matcherData.reduce(
+        (matchersById, matcher) => matchersById.set(matcher.id, matcher), new Map());
+      const errorLineResp = await fetch(getProjectJobUrl('/text_log_errors/', job.id));
+      const errorLineData = await errorLineResp.json();
+      const errorLines = errorLineData.map(line => new ErrorLineData(line))
+        .sort((a, b) => a.data.id - b.data.id);
+      // Lines that are selected
+      let selectedLineIds;
+      let editableLines;
+
+      if (errorLines.length) {
+        const selected = errorLines.find(line => !line.verified);
+        selectedLineIds = new Set([selected.id]);
+        editableLines = errorLines.reduce((pending, line) => (
+          !line.verified ? { ...pending, [line.id]: line } : pending
+        ), {});
+      }
+
+      this.setState({
+        errorLines,
+        errorMatchers,
+        selectedLineIds,
+        editableLines,
+        loadStatus: 'ready'
+      });
+    }
+  }
+
+
+  /**
+   * Test if it is possible to save a specific line.
+   * @param {number} lineId - Line id to test.
+   */
+  canSave(lineId) {
+    const { inputByLine, canClassify } = this.state;
+    const settings = inputByLine.get(lineId);
+
+    if (!canClassify) {
+      return false;
+    }
+    if (!settings) {
+      //This can happen when we are switching jobs
+      return false;
+    }
+    if (settings.type === null) {
+      return false;
+    }
+    if (settings.type === "ignore") {
+      return true;
+    }
+    return !!(settings.classifiedFailureId || settings.bugNumber);
+  }
+
+  /**
+   * Test if it is possible to save all in a list of lines.
+   */
+  canSaveAll() {
+    const pendingLines = this.getPendingLines();
+
+    return (this.state.canClassify && !!pendingLines.length &&
+      pendingLines.every(line => this.canSave(line.id)));
+  }
+
+  /**
+   * Update and mark verified the classification of a list of lines on
+   * the server.
+   * @param {number[]} lines - Lines to test.
+   */
+  save(lines) {
+    if (!Object.keys(lines).length) {
+      return Promise.reject('No lines to save');
+    }
+    const { errorLines } = this.state;
+    const data = Object.values(lines).map(input => ({
+      id: input.id,
+      best_classification: input.classifiedFailureId || null,
+      bug_number: input.bugNumber
+    }));
+
+    this.setState({ loadStatus: "loading" });
+    return this.ThTextLogErrorsModel
+      .verifyMany(data)
+      .then((resp) => {
+        const newErrorLines = resp.data.reduce((newLines, updatedLine) => {
+          const idx = newLines.findIndex(line => line.id === updatedLine.id);
+          newLines[idx] = new ErrorLineData(updatedLine);
+          return newLines;
+        }, errorLines);
+        this.setState({ errorLines: [...newErrorLines], loadStatus: "ready" });
```

Comment:
Just for my education, what's the point of `[...newErrorLines]` here rather than just returning `newErrorLines` directly?